Python/Algorithms/aditya/traversal.py

```diff
@@ -0,0 +1,123 @@
+    def inorder( root ,l): 
+  
+        if root is not None: 
+            Traversal.inorder(root.left,l) 
+            l.append(root.val )
+            Traversal.inorder(root.right,l) 
+        return l 
+    @staticmethod
+    def postorder(root,l): 
+    
+        if root is not None: 
+            Traversal.postorder(root.left, l) 
+            Traversal.postorder(root.right , l) 
+            l.append(root.val )
+        return l
+    @staticmethod
+    def preorder(root , l): 
+    
+        if root is not None: 
+
+            l.append(root.val )
+            Traversal.preorder(root.left , l) 
+            Traversal.preorder(root.right , l) 
+        return l
+    @staticmethod
+    def levelorder(root , l): 
+        if root is None: 
+            return
+        q = deque([])
+        q.append(root) 
+        while len(q):
+            current = q.popleft()
+            l.append(current.val)  
+            if current.left : 
+                q.append(current.left) 
+            if current.right : 
+                q.append(current.right)
+        return l
+
+class TraversalWrapper :
+    def inorder(self):
+        '''
+        return a list with the tree traversed with inorder
+        '''
+
+        return Traversal.inorder(self.root,[])
+        
+    
+    def postorder(self):
+        '''
+        return a list with the tree traversed with postorder
+        '''
+        return Traversal.postorder(self.root , [])
+        
+
+    def preorder(self):
+        '''
+        return a list with the tree traversed with preorder
+        '''
+        return Traversal.preorder(self.root , [])
+        
+
+
+    def levelorder(self):
+        '''
+        return a list with the tree traversed with levelorder
+        '''
+        return Traversal.levelorder(self.root , [])
+        
+
+    def topologicalsort(self):
+        '''
+        return a list with the tree traversed with topologicalsort
+        '''
+        return Traversal.preorder(self.root , [])
+        
+
+    def bfs(self , val):
+        '''
+        searches for the value val using bfs on the tree
+        '''
+        if self.root is None: 
+            return False
+        q = deque([])
+        q.append(self.root) 
+        while len(q):
+            current = q.popleft()
+            if current.val == val :
+                return True   
+            if current.left : 
+                q.append(current.left) 
+            if current.right : 
+                q.append(current.right)
+        return False
+    
+    def dfs(self , val):
+        '''
+        searches for value val  using dfs on the tree
+        '''
+        stack = []
+        stack.append(self.root)
+        while len(stack):
+            current = stack.pop()
+            if current.val == val:
+                return True
+            if current.right:
+                stack.append(current.right)
+            if current.left:
+                stack.append(current.left)
+        return False
```

Comment:
BFS is common to an entire graph and not restricted to a tree. Remove out of tree and make it generic to a rooted graph. Create a Graph DS

Reply:
bfs for a tree is quite different from , bfs for a graph , there are no cycles , no adjacency list , so i dont quite get it , how i can implement same function for the tree and graph ?